Improve type hints in nws

kamiyo opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add return types to methods

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #83058 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

kamiyo committed:
Add return types to methods

bouwew reviewed (commented):
https://github.com/home-assistant/core/blob/be116fe0348a01b38d5e53db826bda8ddecd41b6/homeassistant/components/nws/config_flow.py#L20
Please add typing to this line too, `async def validate_input(hass: core.HomeAssistant, data: dict[str, ?]) -> dict[str, Any]:` Replace `Any` by a fixed format if it is always the same.

bouwew reviewed (commented):
https://github.com/home-assistant/core/blob/be116fe0348a01b38d5e53db826bda8ddecd41b6/homeassistant/components/nws/config_flow.py#L50
Please type errors, could be `errors: dict[str, str] = {}`

bouwew reviewed (commented):
https://github.com/home-assistant/core/blob/be116fe0348a01b38d5e53db826bda8ddecd41b6/homeassistant/components/nws/weather.py#L50

https://github.com/home-assistant/core/blob/be116fe0348a01b38d5e53db826bda8ddecd41b6/homeassistant/components/nws/weather.py#L98

https://github.com/home-assistant/core/blob/be116fe0348a01b38d5e53db826bda8ddecd41b6/homeassistant/components/nws/weather.py#L222

https://github.com/home-assistant/core/blob/be116fe0348a01b38d5e53db826bda8ddecd41b6/homeassistant/components/nws/weather.py#L226

Please add typing to these lines as well.

kamiyo committed:
Add return types to config_flow: validate_input, __init__, and errors dict within async_step_user
Add NWSForecast type that extends Forecast for extra properties
Add return types to weather.py convert_condition, forecast
Add type to const.py: CONDITION_CLASSES

bouwew reviewed (commented):
Great progress!

Still, one line is missing some typing: https://github.com/home-assistant/core/blob/a940ac3191fc298bc48832af514a2d63d0e554f3/homeassistant/components/nws/weather.py#L108

`entry_data`, `hass_data`, `mode` and `units` still need to be typed.

kamiyo commented:
Oh argument types. Got it.

On Mon, Dec 5, 2022, 04:05 Bouwe Westerdijk ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Great progress!
>
> Still, one line is missing some typing:
> https://github.com/home-assistant/core/blob/a940ac3191fc298bc48832af514a2d63d0e554f3/homeassistant/components/nws/weather.py#L108
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/83173#pullrequestreview-1204237156>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWCYT2PE2UCPOJYF3YZKZLWLW47TANCNFSM6AAAAAASSR3K4E>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

kamiyo committed:
Add types for device_info arguments, base_unique_id arguments, in __init__.py
Add types for NWSSensor __init__ arguments in sensor.py
Add types for NWSWeather __init__ arguments in weather.py

bouwew reviewed (approved):
Looking good!

epenet commented on homeassistant/components/nws/__init__.py:
```diff
@@ -37,7 +37,7 @@
 
 
-def base_unique_id(latitude, longitude):
+def base_unique_id(latitude: str, longitude: str):
```
```suggestion
def base_unique_id(latitude: str, longitude: str) -> str:
```

kamiyo committed:
Add return type to base_unique_id in __init__.py
Add typing to async_step_user in config_flow.py

kamiyo committed:
Reorder imports to satisfy isort in sensor.py

MatthewFlamm commented on homeassistant/components/nws/weather.py:
```diff
@@ -205,23 +227,23 @@ def condition(self):
 
     @property
-    def forecast(self):
+    def forecast(self) -> list[Forecast] | None:
```
This line returns type `Forecast` inside the list, but you also created `NWSForecast` which is used for `forecast`.  This seems inconsistent to me.

MatthewFlamm replied:
Putting this back into a thread:

> Yes, sorry my python isn't good enough, but from what I gathered, the weather base class is expecting a Forecast type. But this getter is returning extra properties, which isn't allowed on Forecast since it's a TypedDict. The only way I can get it to pass is to extend Forecast for the internal accumulator, but return the base class in the function. Let me
know if there's a better way.

The current implementation passes the type check. I'm ok keeping the current implementation, or removing `NWSForecast` entirely (it is only for typing).  If we keep the current implementation, it would be good to comment that the `NWSForecast` object is only for typing.  We could also consider explicitly requiring that it is used for typing only using `typing.TYPE_CHECKING`.

kamiyo replied:
I'll mark the object as TYPE_CHECKING. If it were my own project, I'd probably remove the object entirely, but maybe for a future dev it'd be nice to have the hints. I'll push an update now.

kamiyo commented:
Yes, sorry my python isn't good enough, but from what I gathered, the
weather base class is expecting a Forecast type. But this getter is
returning extra properties, which isn't allowed on Forecast since it's a
TypedDict. The only way I can get it to pass is to extend Forecast for the
internal accumulator, but return the base class in the function. Let me
know if there's a better way.

On Thu, Dec 8, 2022, 14:11 MatthewFlamm ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In homeassistant/components/nws/weather.py
> <https://github.com/home-assistant/core/pull/83173#discussion_r1043784457>
> :
>
> >          """Return visibility unit."""
>          return UnitOfLength.METERS
>
>      @property
> -    def forecast(self):
> +    def forecast(self) -> list[Forecast] | None:
>
> This line returns type Forecast inside the list, but you also created
> NWSForecast which is used for forecast. This seems inconsistent to me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/83173#pullrequestreview-1210717828>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABWCYT4TB63C4KNTCVCBZELWMI6IVANCNFSM6AAAAAASSR3K4E>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

gjohansson-ST reviewed (changes requested):
Small comment

kamiyo committed:
Switch lat and long types from str to float

kamiyo committed:
Mark NWSForecast as TYPE_CHECKING only

frenck reviewed (approved):
Thanks, @kamiyo 👍 

../Frenck
